### PR TITLE
Bloomberg PR: Non-telemetry improvements

### PR DESCRIPTION
**Pull request by lornasong**
This PR includes the non-telemetry changes from #56, which include a number of scaling, logging, refactoring improvements and features. A separate PR will be open for the telemetry-only changes.

#56 changes were merged into a non-master branch `bloomberg-dev` in order to help carry the PR, add some additional commits to Bloomberg’s work, and retain the Corporate CLA.

### Features in this PR:

**Performance/scaling improvements**

A. Prevents checks from running more than once per second: update check intervals that are less than the minimum interval (1s)
 - `agent.go`: `minimumInterval`, `watchHealthChecks()`
 - `check.go`: `NewCheckRunner()`, `UpdateChecks()`

B. Prevent spurious node status updates: only status on change or expiration
 - `config.go`: `NodeHealthRefreshInterval`
 - `agent.go`: `lastKnownStatus{}`, `Agent.knownNodeStatuses`, `Agent.knownNodeStatusesLock`, `shouldUpdateNodeStatus()`, `updateLastKnownNodeStatus()`
 - `coordinate.go`: `updateHealthyNode()`, `updateHealthyNodeTxn()`, `updateFailedNode()`, `updateFailedNodeTxn()`

C. Support over 64 ESM instances: add transaction handling which has a 64-operation limit
 - `agent.go`: `maximumTransactionSize`
 - `leader.go`: `computeWatchedNodes` consumes `maximumTransactionSize`

**Node coordinate improvements**

D. Option to disable updating node coordinates: new configuration options
 - `readme`: add new config to documentation
 - `config.go`: `DisableCoordinateUpdates`
 - `coordinate.go`: `updateNodeCoordinate()`
 - `main.go`: log out `DisableCoordinateUpdates`

E. Skip updating small changes in node coordinates: update coordinate catalog if it doesn’t exist of the change is greater than a millisecond
 - `coordinate.go`: `updateNodeCoordinate()`

**Refactoring**

F. `leader.go`: refactor out `firstRun` and use `retryTimer`

G. `check.go`: refactor HTTP and TCP check code out of `UpdateChecks` and into `updateCheckTCP` and `updateCheckHTTP`. This also resolves an issue where an ESM follower will operate with stale information when it is unassigned and then reassigned the same check - more details in https://github.com/hashicorp/consul-esm/pull/63#issuecomment-635914358.

**Logging**

H. Additional logs across files

I. Log when checks are added/updated/deleted: in refactoring HTTP and TCP checks (G), sort out checks by add/update/delete. This helps provide visibility into resolving the issue fixed by G.
 - `check.go`: declare `added`, `updated`, `removed` in `UpdateChecks` and consume them in `updateCheckHTTP`, `updateCheckTCP`, `UpdateChecks`

**Additional**

~~J. Update to Go v 1.13~~ (Moved to https://github.com/hashicorp/consul-esm/pull/64)
~~- `go.mod`~~

### Documentation of Changes since #56

The below roughly align with the commits on May 27 & June 8

**Nits/Minor Changes**
 - Formatting and typos
 - Resolving some linting issues
 - Renamed some variables
 - Updated some comments
 - Some refactoring to avoid else-block
 - Formatting to clean up duplicate typings to shorten method signatures
 - Refactor `time.Now()` out of `isExpired`

**Additions**
 - Added `disable_coordinate_updates` config to Readme
 - Unit tests

**Notable**
 - Removed `DisableRedundantStatusUpdates` from config
 - Fix potential bug when leader commit operation fails

### CODE REVIEW

**Comment by lornasong** [comment]
@mikeyyuen, I am double-checking with our legal team regarding the CLA. In the meantime, this PR is Bloomberg’s non-telemetry changes + some additional commits from us.

"Features in this PR" section in the PR description is the improvements as I understand them extrapolated from #56’s PR description. More just an FYI for future readers, not necessary for you to read in-depth! The only thing I didn't quite pinpoint was the code for "Fix an issue where check updates can fail to propagate to ESM Followers", so let me know where I’ve missed that.

"Documentation of Changes since #56" has an overview of the additional commits made. I've roughly broken the commits out to match the overview for easy reviewing.

Most notable changes that I especially would like your feedback on are:
 - Removed `DisableRedundantStatusUpdates` https://github.com/hashicorp/consul-esm/pull/63/commits/bc4b7892b76cd76c1fcb77378faee01b1b29c1cd
 - Fix an issue when leader commit operation fails: https://github.com/hashicorp/consul-esm/pull/63/commits/a0e438a32536342d295852292a6506188e2b8131

Let me know anything I misunderstood, and see the commit message for details. Don’t hesitate to share any other feedback on the other commits too.

I will also have other members of my team take a look and review as well so there may be additional commits.

Let me know if you have questions. Thanks!

**Comment by mikeyyuen** [comment]
Re: DisableRedundantStatusUpdates

Yes, your right, I just checked our internal copy of the source and it isn't used. I think it was intended to allow users to disable this distance check in coordinate.go

```
if len(coords) == 0 || coord.Coord.DistanceTo(newCoord) > time.Millisecond {
```

We didn't need to disable it in the end. If your happy for this to be the default behavior then the config option isn't needed.

**Comment by mikeyyuen** [comment]
Re: leader commit operation fails

Yes, that makes sense! Thank you for patching it =]

**Comment by mikeyyuen** [comment]
Re: Fix an issue where check updates can fail to propagate to ESM Followers

From memory & looking at the code, I think this was a nasty one and it relates to all the agent.go changes.

When the ESM leader would move responsibility for a node from ESM follower (a) to another ESM follower (b). If the if (b) had already seen that node from a previous period where it was responsible for it, then it would have stale state still in memory for that node. It would make any updates based on that stale state, this would cause checks to not be performed.

I think its not obvious from the agent.go change as we had to do a bit of a refactor to get it to work, but the way we now don't refer back the the definition when deleting the `checksHTTP` and `checksTCP` I think was the fix.

Sorry if thats a bit vague, It was over a year ago that we found that one. I remember it took a while and is why the added / removed / updated debug was put in.

**Comment by lornasong** [comment]
@mikeyyuen,  thanks for taking a look and the feedback!

Re: DisableRedundantStatusUpdates - helpful to know. I think this default behavior makes sense. I’ll leave it as is and add the configuration if we hear feedback from our community.

Re: Fix an issue where check updates can fail to propagate to ESM Followers - thanks for the context. What you’re saying makes a lot of sense to me. If I understand you correctly, I noticed that you replaced some of the `if c.checks[checkHash].Definition.HTTP != ""` with `if httpCheck, httpCheckExists := c.checksHTTP[checkHash]; httpCheckExists`, which I had found more readable. Now that you point it out, I can imagine a scenario like you describe where the check on the `Definition` could be problematic. I’ll add your details to the PR description.

Just to be transparent on what I’m planning to do next:
 - CLA: waiting to hear back from legal and will go from there as far as logistics on getting the improvements merged into master.
 - Team review: reach out to my team next week to request feedback.

As always, if you think of any feedback on the additional commits, don’t hesitate to share.

Let me know if there’s any other updates I can provide. Thanks!

File: go.mod [hunk 1]
```diff
@@ -1,16 +1,30 @@
 module github.com/hashicorp/consul-esm
 
-go 1.12
+go 1.13
```
[the remark this answers is not in the document]
Comment (eikenb) [reply go.mod:3]: +1. This should be split out. Particularly the dependency updates. Those should always be in a separate merge.

File: check.go [hunk 2]
```diff
@@ -67,13 +76,117 @@ func (c *CheckRunner) Stop() {
 	}
 }
 
+// Update an HTTP check
+func (c *CheckRunner) updateCheckHTTP(latestCheck *api.HealthCheck, checkHash types.CheckID, definition *api.HealthCheckDefinition, updated checkIDSet, added checkIDSet) bool {
```
Comment (eikenb) [inline check.go:80]: Not going to require this to be fixed, but a 178 character code line is way to long.

**Comment by eikenb** [comment]
I looked it over and it looks generally OK. I think some of the lines are too long and I think it should have been broken up into separate commits/PRs. I don't think either of these is enough to require a reworking before merging at this point, but I think we should ask for either separate PRs or separate commits for each of the changesets in the future.

File: check.go (hunk 2 above)
Comment (lornasong) [reply check.go:80]: Thanks for this feedback! Yes, I think we could improve readability here. @findkim mentioned this to me as well suggesting I could clean up duplicate types. I'll do that and add a newline after `checkHash types.CheckID`. I'll do this for `updateCheckTCP` as well and something similar for `NewCheckRunner`

File: coordinate.go [hunk 3]
```diff
@@ -315,18 +358,24 @@ func (a *Agent) updateNodeCoordinate(node *api.Node, rtt time.Duration) error {
...
-	// Update the coordinate in the catalog.
-	_, err = a.client.Coordinate().Update(&api.CoordinateEntry{
-		Node:    coord.Node,
-		Segment: coord.Segment,
-		Coord:   newCoord,
-	}, nil)
+	// Update the coordinate in the catalog
+	// - if there is a significant change
+	// - if there isn't a coordinate already there
+	if len(coords) == 0 || coord.Coord.DistanceTo(newCoord) > time.Millisecond {
```
[the remark this answers is not in the document]
Comment (lornasong) [reply coordinate.go]: I like this! Big fan of sparing indentation.

File: go.mod (hunk 1 above)
[the remark this answers is not in the document]
Comment (lornasong) [reply go.mod]: Great point, thank you! Linking the PR for updating go version: https://github.com/hashicorp/consul-esm/pull/64

**Comment by lornasong** [comment]
@findkim and @eikenb - thanks so much for the review. I've worked on the feedback and this is ready for re-review when you have a chance. Let me know any additional feedback!

@findkim - I've additionally added some changes around refactoring `isExpired` and `TestAgent_shouldUpdateNodeStatus` that we discussed offline. Also created https://github.com/hashicorp/consul-esm/issues/65 for the `Config` vs. `HumanConfig` issue we discussed.

@eikenb - I think that’s a great point about requesting separate commit/PR per changeset. I was struggling to break this into further PRs since some features cross multiple files but I think I could have definitely split by commit. I’ll keep this in mind and make sure to push for this in the future.

**Comment by lornasong** [comment]
After checking with legal, I will go ahead and merge this though the CLA appears pending on this PR. This new code originates from Bloomberg, with several changes by our team prior to merge. Bloomberg has an existing Corporate CLA.

**Pull request merged**